### PR TITLE
Space varying material properties

### PR DESCRIPTION
**Pull request by dalg24**

### CODE REVIEW

File: cpp/source/deal.II/mp_values.h [hunk 1]
```diff
@@ -101,6 +93,26 @@ class UniformConstantMPValues : public MPValues<dim>
...
+    \note
+    Here \p fe_values must be constructed with the flag \c
+    dealii::update_quadrature_points.
+  */
+  void get_values(std::string const &key,
+                  dealii::FEValues<dim> const &fe_values,
+                  std::vector<double> &values) const override;
+
+private:
+  std::shared_ptr<dealii::Function<dim>> const _function;
```
Comment (Rombur) [inline cpp/source/deal.II/mp_values.h]: Why do you use a Function and not FunctionParser?

File: cpp/source/utils.templates.h [hunk 2]
```diff
@@ -8,12 +8,25 @@
...
 #include <boost/lexical_cast.hpp>
 #include <boost/algorithm/string.hpp>
+#include <boost/regex.hpp>
 #include <boost/foreach.hpp>
 #include <type_traits>
 
 namespace cap
 {
 
+namespace impl
```
Comment (Rombur) [inline cpp/source/utils.templates.h]: What not `namespace internal`? Do we need to have a different namespace?

File: cpp/source/utils.templates.h [hunk 3]
```diff
@@ -47,4 +54,28 @@ std::string to_string(std::vector<T> const &v)
...
+  std::vector<std::string> pairs;
+  boost::algorithm::split(pairs, s, boost::algorithm::is_any_of(","));
+  for (auto const &p : pairs)
+  {
+    std::vector<std::string> kv;
+    boost::algorithm::split(kv, p, boost::algorithm::is_any_of("="));
+    if (kv.size() != 2)
+      throw std::runtime_error("invalid key-value pair " + p);
+    for (auto &x : kv)
+      boost::algorithm::trim(x);
```
Comment (Rombur) [inline cpp/source/utils.templates.h:73]: Does this get ride of blank space?

File: cpp/test/test_mp_values.cc [hunk 4]
```diff
@@ -14,9 +14,85 @@
...
+  dealii::Triangulation<dim> triangulation;
+  dealii::GridGenerator::hyper_cube(triangulation);
+  dealii::DoFHandler<dim> dof_handler(triangulation);
+  dealii::DoFHandler<dim>::active_cell_iterator cell =
+      dof_handler.begin_active();
+
+  std::vector<dealii::Point<dim>> points = {{0, 0}, {0, 1}, {0.5, 0.5}, {1, 1}};
+  // NOTE: I tried with dealii::update_default but I got a memory access
+  // violation instead of some deal.II exception as I expected...
+  // I was using deal.II in release mode.
```
Comment (Rombur) [inline cpp/test/test_mp_values.cc:67]: There are very few checks in release mode. That's why it goes at least five times faster.

File: cpp/source/utils.templates.h (hunk 3 above)
Comment (dalg24) [reply cpp/source/utils.templates.h:73]: strip whitespace left and right

File: cpp/source/utils.templates.h (hunk 2 above)
Comment (dalg24) [reply cpp/source/utils.templates.h]: k will change it

File: cpp/source/deal.II/mp_values.h (hunk 1 above)
Comment (dalg24) [reply cpp/source/deal.II/mp_values.h]: more general.  we could consider move uniform constant to simple `dealii::ConstantFunction<dim>`

**Comment by dalg24** [comment]
@Rombur I changed the namespace and committed the switch that allows us to overwrite \kappa in the electrode and the separator.  All we need is adding a node `custom_liquid_electrical_conductivity { expression: ..., variables[optional]: ..., constants[optional]: ... }` in `material_properties.{electrode,separator}_material`.

**Comment by Rombur** [comment]
tests don't pass

**Comment by dalg24** [comment]
ping.  in particular, look at c7a4051.

**Comment by dalg24** [comment]
only the "timer" test failed so I'd say this is all good

File: cpp/source/deal.II/mp_values.h [hunk 5]
```diff
@@ -101,6 +93,26 @@ class UniformConstantMPValues : public MPValues<dim>
 };
 
 template <int dim>
+class FunctionSpaceMPValues : public MPValues<dim>
+{
+public:
+  FunctionSpaceMPValues(std::shared_ptr<dealii::Function<dim>> const &function);
```
Comment (Rombur) [inline cpp/source/deal.II/mp_values.h]: `const &`? Also add blank line between the two constructors.

File: cpp/source/utils.templates.h [hunk 6]
```diff
@@ -8,12 +8,25 @@
...
 
+namespace internal
+{
+void filter_boolean(std::string &b)
+{
+  boost::algorithm::to_lower(b);
+  if (b.compare("true") == 0)
+    b = "1";
+  else if (b.compare("false") == 0)
+    b = "0";
```
Comment (Rombur) [inline cpp/source/utils.templates.h:26]: This should throw an exception in the `else` case.

**Review by Rombur — APPROVED** [review]
Looks good please fix the two small comments and merge yourself.

File: cpp/source/utils.templates.h (hunk 6 above)
Comment (dalg24) [reply cpp/source/utils.templates.h:26]: Not necessary, the `boost::lexical_cast<boost>()` will throw otherwhise

File: cpp/source/deal.II/mp_values.h (hunk 5 above)
Comment (dalg24) [reply cpp/source/deal.II/mp_values.h]: Yeah, the `shared_ptr` is still an object;)  You want me to pass by value?

**Pull request merged**